Snakefile

```diff
@@ -183,49 +184,154 @@ rule get_sequences_without_nextclade_annotations:
         """
 
-rule annotate_via_nextclade:
+GENES = "E,M,N,ORF1a,ORF1b,ORF3a,ORF6,ORF7a,ORF7b,ORF8,ORF9b,S"
+GENES_SPACE_DELIMITED = GENES.replace(",", " ")
+
+rule run_nextclade:
     message:
         """
-        If there are sequences without clades, then run nextclade to assign
-        clades and calculate other useful metrics.
+        Runs nextclade on sequences which were not in the previously cached nextclade run.
+        This alignes sequences, assigns clades and calculates some of the other useful
+        metrics which will ultimately end up in metadata.tsv.
         """
     input:
-        sequences = f"data/{database}/nextclade.sequences.fasta",
-        nextclade_info = f"data/{database}/nextclade_old.tsv"
+        sequences = f"data/{database}/nextclade.sequences.fasta"
     params:
-        new_info = temp(f"data/{database}/nextclade_new.tsv"),
         nextclade_input_dir = temp(directory(f"data/{database}/nextclade_inputs")),
-        nextclade_output_dir = temp(directory(f"data/{database}/nextclade"))
-    threads: 16
+        nextclade_output_dir = temp(directory(f"data/{database}/nextclade")),
+    threads: 64
+    output:
+        info = f"data/{database}/nextclade_new.tsv",
+        alignment = temp(f"data/{database}/nextclade.aligned.upd.fasta"),
+        insertions = temp(f"data/{database}/nextclade.insertions.csv")
+    shell:
+        """
+        ./bin/run-nextclade \
+            {input.sequences:q} \
+            {output.info} \
+            {params.nextclade_input_dir} \
+            {params.nextclade_output_dir} \
+            {output.alignment} \
+            {output.insertions} \
+            {GENES} \
+            {threads}
+        """
+
+rule nextclade_info:
+    message:
+        """
+        Generates nextclade info TSV for all sequences (new + old)
+        """
+    input:
+        old_info = f"data/{database}/nextclade_old.tsv",
+        new_info = f"data/{database}/nextclade_new.tsv"
     output:
         nextclade_info = f"data/{database}/nextclade.tsv"
+    shell:
+        """
+        ./bin/join-rows \
+            {input.old_info:q} \
+            {input.new_info:q} \
+            -o {output.nextclade_info:q}
+        """
+
+rule download_previous_alignment:
+    ## NOTE two potential bugs with this implementation:
+    ## (1) race condition. This file may be updated on the remote after download_nextclade has run but before this rule
+    ## (2) we may get `download_nextclade` and `download_previous_alignment` from different s3 buckets
+    params:
+        dst_source = config["s3_dst"] + '/nextclade.aligned.fasta.xz',
+        src_source = config["s3_src"] + '/nextclade.aligned.fasta.xz',
+    output:
+        alignment = temp(f"data/{database}/nextclade.aligned.old.fasta")
+    shell:
+        """
+        ./bin/download-from-s3 {params.dst_source} {output.alignment} ||  \
+        ./bin/download-from-s3 {params.src_source} {output.alignment}
+        """
```

Comment:
Right now I am only concerned by this comment from James.
@jameshadfield Can you tell a little more about that?

> This file may be updated on the remote after download_nextclade has run but before this rule

So this means someone else would update the alignment on S3?

The relevant steps in the workflow are (roughly ordered):

 1. download from S3 the old nextclade.tsv
 2. start running Nextclade
 3. finish running Nextclade
 4. concatenate old and new nextclade.tsv
 5. upload nextclade.tsv
 6. download old aligned.fasta
 7. concatenate old and new aligned.fasta.tsv
 8. upload aligned.fasta

The steps 4-5 run concurrently with 6-8 and 6 runs concurrently with everything else, including 1. So if someone updates aligned.fasta on S3 after 1, but before 6, then the old nextclade.tsv is no longer in sync with old aligned.fasta. In fact, there are many more permutations in which things can go wrong.

The hope is that if someone updates the alignment.fasta, they'd also update the nextclade.tsv ***at the same time*** and won't update anything while daily ingest is running. Otherwise it can happen that the newly uploaded alignment.fasta and nextclade.tsv are overwritten by the workflow and the new data (e.g. the new clades from the full run) will be lost.

But I cannot think of who is that bad guy who silently updates S3 is. It could be Nextclade full run, but it produces files into a separate directory and then, after a manual verification, the copy to the "master" location is made by hand. So whoever does the copy need to make sure they do this at the right moment. But this was always true.

Everyone with write access to S3 should be aware of that and we need a "social mutex" mechanism to ensure consistency of operations :)


Again, the entire problem stems from the fact that we are using Snakemake, which introduces concurrency that we don't need. Is there a way to suppress it somehow?

(And another fundamental problem of course is that we treat a database a a set of files, and basically reinventing an RDBMS in a really shitty and slow way... We already have filtering, joins, etc. so now we need transactions, I guess)

Reply:
> Everyone with write access to S3 should be aware of that and we need a "social mutex" mechanism to ensure consistency of operations :)

Yes -- I think this race condition will rarely, if ever, be hit. I only expect one action to modify these files (the re-run full nextclade action).

> Again, the entire problem stems from the fact that we are using Snakemake

This also existed in the non-snakemake version.